go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go

```diff
@@ -246,6 +246,267 @@ func TestPlayerCopyBigTable(t *testing.T) {
+		// Sleeping 50ms guarantees that the catchup wait loop executes multiple times.
+		// This is because waitRetryTime is set to 10ms.
+		time.Sleep(50 * time.Millisecond)
+		// Do this no more than once.
+		vstreamRowsHook = nil
+	}
+
+	filter := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match:  "/.*",
+			Filter: "",
+		}},
+	}
+
+	bls := &binlogdatapb.BinlogSource{
+		Keyspace: env.KeyspaceName,
+		Shard:    env.ShardName,
+		Filter:   filter,
+		OnDdl:    binlogdatapb.OnDDLAction_IGNORE,
+	}
+	query := binlogplayer.CreateVReplicationState("test", bls, "", binlogplayer.VReplicationInit, playerEngine.dbName)
+	qr, err := playerEngine.Exec(query)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		query := fmt.Sprintf("delete from _vt.vreplication where id = %d", qr.InsertID)
+		if _, err := playerEngine.Exec(query); err != nil {
+			t.Fatal(err)
+		}
+		expectDBClientQueries(t, []string{
+			"/delete",
+		})
+	}()
+
+	expectDBClientQueries(t, []string{
+		"/insert into _vt.vreplication",
+		// Create the list of tables to copy and transition to Copying state.
+		"begin",
+		"/insert into _vt.copy_state",
+		"/update _vt.vreplication set state='Copying'",
+		"commit",
+		"rollback",
+		// The first fast-forward has no starting point. So, it just saves the current position.
+		"/update _vt.vreplication set pos=",
+		"begin",
+		"insert into src(id,val) values (1,'aaa')",
+		`/update _vt.copy_state set lastpk='fields:<name:\\"id\\" type:INT32 > rows:<lengths:1 values:\\"1\\" > ' where vrepl_id=.*`,
+		"commit",
+		"rollback",
+		// The next catchup executes the new row insert, but will be a no-op.
+		"begin",
+		"insert into src(id,val) select 3, 'ccc' from dual where (3) <= (1)",
+		"/update _vt.vreplication set pos=",
+		"commit",
+		// fastForward has nothing to add. Just saves position.
+		"begin",
+		"/update _vt.vreplication set pos=",
+		"commit",
+		// Second row gets copied.
+		"begin",
+		"insert into src(id,val) values (2,'bbb')",
+		`/update _vt.copy_state set lastpk='fields:<name:\\"id\\" type:INT32 > rows:<lengths:1 values:\\"2\\" > ' where vrepl_id=.*`,
+		"commit",
+		// Third row copied without going back to catchup state.
+		"begin",
+		"insert into src(id,val) values (3,'ccc')",
+		`/update _vt.copy_state set lastpk='fields:<name:\\"id\\" type:INT32 > rows:<lengths:1 values:\\"3\\" > ' where vrepl_id=.*`,
+		"commit",
+		"/delete from _vt.copy_state.*src",
+		// rollback is a no-op because the delete is autocommitted.
+		"rollback",
+		// Copy is done. Go into running state.
+		"/update _vt.vreplication set state='Running'",
+		// All tables copied. Final catch up followed by Running state.
+	})
+	expectData(t, "src", [][]string{
+		{"1", "aaa"},
+		{"2", "bbb"},
+		{"3", "ccc"},
+	})
+}
+
+// TestPlayerCopyJsonData ensures the copy-catchup back-and-forth loop works correctly
+// when the filter uses a wildcard rule & the tables have json data
+func TestPlayerCopyJsonData(t *testing.T) {
```

Comment:
We should split this into two tests, one for copy and the other for vreplication. The back-and-forth part doesn't need to be combined with this.

Reply:
Done.